### PR TITLE
[stacked 1/3, multi-datasets] Adding most of the multi-dataset cleanly

### PR DESCRIPTION
**Pull request by coreyjadams**
# PhysicsNeMo Pull Request

This pr adds multiple datasets to the physicsnemo datapipe scheme.

They are added dynamically at creation, viewed as one logical dataset, and if passed with a sampler for random access you the length is tracked via the sampler, not the datasets - this is to let you squish a bunch of data together, and then randomly take 80/20 splits of the whole thing, for example.

Transforms are _per dataset_, not uniform.  This deliberately lets you have unique pipelines for unique data.  You can have them come out all funky at the end, if you have batch size 1 and don't need to collate data.  If you _do_ want everything happy and collatable, there is a helper debug setting in the MultiDataset class that will take the first item from every dataset at construction, run the pipeline, and at least verify the keys match.  I can't align the shapes or anything without a collation function too, and since that can vary per dataset, we have to defer.  But it's something.

There are more PRs coming in this stack but it's difficult still to pile them on top with the tools I'm using.  Next up:
- some transform work to clean up the pipelines for an example.
- A new exmaple in the darcy flow space enabling the transolver example with multiple datasets at once.

There is a parallel pr for geotransolver to enable it for 2D data.  With that in, we can do GeoTransolver on Darcy with multiple datasets too.

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: physicsnemo/datapipes/multi_dataset.py [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+        if index < 0 or index >= n:
+            raise IndexError(
+                f"Index {index} out of range for MultiDataset with {n} samples"
+            )
+        # Find k such that cumul[k] <= index < cumul[k+1]
+        for k in range(len(self._cumul) - 1):
+            if self._cumul[k] <= index < self._cumul[k + 1]:
+                return k, index - self._cumul[k]
+        # Fallback (should not be reached)
+        return len(self._datasets) - 1, index - self._cumul[-2]
```
[the remark this answers is not in the document]
Comment (peterdsharpe) [reply physicsnemo/datapipes/multi_dataset.py:219]: Good bot! 🤖

File: physicsnemo/datapipes/multi_dataset.py [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+    >>> multi = MultiDataset([ds_a, ds_b], output_strict=True)                   # doctest: +SKIP
+    >>> len(multi) == len(ds_a) + len(ds_b)                                      # doctest: +SKIP
+    True
+    >>> data, meta = multi[0]   # from ds_a                                      # doctest: +SKIP
+    >>> meta["dataset_index"]   # 0                                              # doctest: +SKIP
+    """
+
+    def __init__(
+        self,
+        datasets: Sequence[Dataset],
```
Comment (laserkelvin) [inline physicsnemo/datapipes/multi_dataset.py]: This is purely stylistic, but I would prefer `*datasets: Dataset` so users don't have to form a list, or if they do, can just unpack.

File: physicsnemo/datapipes/multi_dataset.py [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+
+        # Cumulative lengths: cumul[k] = sum(len(datasets[j]) for j in range(k))
+        # So index i is in dataset k when cumul[k] <= i < cumul[k+1], local = i - cumul[k]
+        cumul = [0]
+        for ds in self._datasets:
+            cumul.append(cumul[-1] + len(ds))
+        self._cumul = cumul
+
+        if output_strict:
+            self._field_names = _validate_strict_outputs(self._datasets)
```
Comment (laserkelvin) [inline physicsnemo/datapipes/multi_dataset.py:184]: While keys are validated, what about shapes and data types?

File: docs/api/datapipes/physicsnemo.datapipes.rst [hunk 4]
```diff
@@ -153,35 +155,60 @@ the ``Dataset`` is responsible for the threaded execution of ``Reader``s and
...
-container chains multiple transforms into a single callable.  See
+container chains multiple transforms into a single callable.  Refer to
 :doc:`physicsnemo.datapipes.transforms` for the base class API, ``Compose``,
 and all built-in transforms.
 
 Collation
 ^^^^^^^^^
 
-Combining a set of tensordict objects into a batch of data can, at times,
+Combining a set of TensorDict objects into a batch of data can, at times,
```
Comment (megnvidia) [inline docs/api/datapipes/physicsnemo.datapipes.rst]: This looked like one of those I was typing some text and then a bunch of notifications hit my screen and I lost track of where my cursor was and I thought i was done with what i had written in the one window....and then someone walked by and started talking to me about something.......

I took a guess at what a logical way to complete the sentence was.

File: physicsnemo/datapipes/multi_dataset.py (hunk 2 above)
Comment (coreyjadams) [reply physicsnemo/datapipes/multi_dataset.py]: Good call.

File: physicsnemo/datapipes/multi_dataset.py (hunk 3 above)
Comment (coreyjadams) [reply physicsnemo/datapipes/multi_dataset.py:184]: Shapes and data types are not validated.  data types I could add, there are not really any scenarios I can think of where batches are added together and mixing of dtypes happens.

Shape compatibility depends entirely on what collation function is used, and that is not exposed to the dataset.  Only to the DataLoader.  So, I can't validate it here.  If we really need that validation, I think we encourage users to try using the whole pipeline?

**Comment by coreyjadams** [comment]
@laserkelvin thanks for taking a look!

- yes, there is a line of sight towards DDP.  #1507 includes this PR + a training example.  I didn't build DDP into it, but can (and now will!) but tl;dr it's not hard;  we use the usual machinery of distributed samplers to coordinate over ranks and it works as usual.  Datasets are exposed as "one" unit so the sampler indexes into the unit as a whole.
- Users should expect to balance batch composition via samplers, if they want that behavior.  The idea here was that if you compose datasets into a multi-dataset, they become "one" and your sampling treats them as one.  Perhaps you think we should support some sort of custom sampler?  We run the risk of biasing the training of course in that case...
- the intention is not solely for concatenation.  In fact, for many datasets collation via concatenation is not correct.  So I can't do something more sophisticated than check "hey, does every dataset produce the same keys?"  It is not strictly mandatory, though, you could have a different output from each dataset happily, if you are not collating (aka batch size 1).
- yes

File: docs/api/datapipes/physicsnemo.datapipes.rst (hunk 4 above)
Comment (coreyjadams) [reply docs/api/datapipes/physicsnemo.datapipes.rst]: Ah, it was just a small typo needing to join up with the next sentence.  I'll fix it.  Thanks!

**Comment by laserkelvin** [comment]
> * Users should expect to balance batch composition via samplers, if they want that behavior.  The idea here was that if you compose datasets into a multi-dataset, they become "one" and your sampling treats them as one.  Perhaps you think we should support some sort of custom sampler?  We run the risk of biasing the training of course in that case...

Yeah nah I don't think we need to necessary provide the custom sampler - that will depend a lot on the dataset compositions, and we can't know that _a priori_ generally, I don't think. **But**, I think it is important to highlight that aspect, maybe in an example or something when it comes to it. I imagine there's going to be a lot of imbalanced datasets that, if naively just shuffled, will end up yielding training signals skewed towards the bigger datasets. As long as there's a path forward for then I'm happy with that.

Another thing came to mind as well: how do you envision dealing with different splits?
